sqlx-core/src/query.rs

```diff
@@ -78,13 +83,29 @@ impl<'q, DB: Database> Query<'q, DB, <DB as Database>::Arguments<'q>> {
     ///
     /// There is no validation that the value is of the type expected by the query. Most SQL
     /// flavors will perform type coercion (Postgres will return a database error).
+    ///
+    /// If encoding the value fails, the error is stored and later surfaced when executing the query.
     pub fn bind<T: 'q + Encode<'q, DB> + Type<DB>>(mut self, value: T) -> Self {
-        if let Some(arguments) = &mut self.arguments {
-            arguments.add(value);
+        if let Err(error) = self.try_bind(value) {
+            self.arguments = Some(Err(error));
```

Comment:
Thinking from the perspective of someone debugging this, it would be really useful if this stored the number of times `bind()` was called previously so the user could find exactly _which_ `bind()` call is failing in their code if they have several of them.

Reply:
I've solved this by adding a `len()` method to the `Arguments` trait. This way `Query::bind` can just ask the arguments how many there already are and report the position accordingly instead of having to track it separately.